doc/bspwm.1
'\" t
.\"     Title: bspwm
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 10/01/2013
.\"    Manual: Bspwm Manual
.\"    Source: Bspwm 0.8.5
.\"  Language: English
.\"
.TH "BSPWM" "1" "10/01/2013" "Bspwm 0\&.8\&.5" "Bspwm Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
bspwm \- Binary space partitioning window manager
.SH "SYNOPSIS"
.sp
\fBbspwm\fR [\fB\-h\fR|\fB\-v\fR|\fB\-c\fR \fICONFIG_PATH\fR|\fB\-s\fR \fIPANEL_FIFO\fR|\fB\-p\fR \fIPANEL_PREFIX\fR]
.sp
\fBbspc\fR \fICOMMAND\fR [\fIARGUMENTS\fR]
.SH "DESCRIPTION"
.sp
\fBbspwm\fR is a tiling window manager that represents windows as the leaves of a full binary tree\&.
.sp
It is controlled and configured via \fBbspc\fR\&.
.SH "OPTIONS"
.PP
\fB\-h\fR
.RS 4
Print the synopsis and exit\&.
.RE
.PP
\fB\-v\fR
.RS 4
Print the version and exit\&.
.RE
.PP
\fB\-c\fR \fICONFIG_PATH\fR
.RS 4
Use the given configuration file\&.
.RE
.PP
\fB\-s\fR \fIPANEL_FIFO\fR
.RS 4
Write the internal state to the given FIFO\&.
.RE
.PP
\fB\-p\fR \fIPANEL_PREFIX\fR
.RS 4
Start every line written to the
\fIPANEL_FIFO\fR
with the given prefix\&.
.RE
.SH "CONFIGURATION"
.sp
\fBbspwm\fR has only two sources of informations: the X events it receives and the messages it reads on a dedicated socket\&.
.sp
The default configuration file is \fI$XDG_CONFIG_HOME/bspwm/bspwmrc\fR\&.
.sp
Keyboard and pointer bindings are defined with sxhkd\&.
.sp
Example configuration files can be found in the \fBexamples\fR directory\&.
.SH "SPLITTING MODES"
.sp
New windows are inserted in the tree as close as possible to the focused window\&.
.sp
There is only two splitting modes: \fIautomatic\fR and \fImanual\fR\&.
.sp
The default mode is \fIautomatic\fR\&. The \fImanual\fR mode is entered by sending a \fBpreselection\fR message\&.
.sp
Example: insertion of a new node (number 4) into the given tree in \fIautomatic\fR mode:
.sp
.if n \{\
.RS 4
.\}
.nf
              b                                   c
             / \e                                 / \e
            3   a              \-\->              4   b
            ^  / \e                              ^  / \e
              2   1                               3   a
                                                     / \e
                                                    2   1
 +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+         +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
 |            |            |         |            |            |
 |            |     2      |         |            |     3      |
 |            |            |         |            |            |
 |     3      |\-\-\-\-\-\-\-\-\-\-\-\-|   \-\->   |     4      |\-\-\-\-\-\-\-\-\-\-\-\-|
 |     ^      |            |         |     ^      |     |      |
 |            |     1      |         |            |  1  |  2   |
 |            |            |         |            |     |      |
 +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+         +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
.fi
.if n \{\
.RE
.\}
.sp
Same departure, but the mode is \fImanual\fR, and a \fBwindow \-\-presel up\fR message was sent beforehand:
.sp
.if n \{\
.RS 4
.\}
.nf
             b                                   b
            / \e                                 / \e
           3   a              \-\->              c   a
           ^  / \e                             / \e / \e
             2   1                           4  3 2  1
                                             ^
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+         +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
|            |            |         |            |            |
|            |     2      |         |     4      |     2      |
|            |            |         |     ^      |            |
|     3      |\-\-\-\-\-\-\-\-\-\-\-\-|   \-\->   |\-\-\-\-\-\-\-\-\-\-\-\-|\-\-\-\-\-\-\-\-\-\-\-\-|
|     ^      |            |         |            |            |
|            |     1      |         |     3      |     1      |
|            |            |         |            |            |
+\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+         +\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-+
.fi
.if n \{\
.RE
.\}
.SH "CONTAINERS"
.sp
Each monitor contains at least one desktop\&.
.sp
Each desktop contains at most one tree\&.
.SH "COMMON DEFINITIONS"
.sp
.if n \{\
.RS 4
.\}
.nf
DIR         := left | right | up | down
CYCLE_DIR   := next | prev
.fi
.if n \{\
.RE
.\}
.SH "SELECTORS"
.sp
Selectors are used to select a target window, desktop, or monitor\&. A selector can either describe the target relatively or name it globally\&.
.sp
Descriptive (relative) selectors consist of a primary selector and any number of non\-conflicting modifiers as follows:
.sp
.if n \{\
.RS 4
.\}
.nf
PRIMARY_SELECTOR[\&.MODIFIER]*
.fi
.if n \{\
.RE
.\}
.sp
For obvious reasons, neither desktops nor monitors names may be valid descriptive selectors\&.
.SS "Window"
.sp
Select a window\&.
.sp
.if n \{\
.RS 4
.\}
.nf
WINDOW_SEL := <window_id>
            | (DIR|CYCLE_DIR|biggest|last|focused)[\&.floating|\&.tiled][\&.like|\&.unlike][\&.automatic|\&.manual][\&.urgent|\&.nonurgent]
.fi
.if n \{\
.RE
.\}
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBPrimary Selectors\fR
.RS 4
.PP
\fIDIR\fR
.RS 4
Selects the window in the given (spacial) direction relative to the active window\&.
.RE
.PP
\fICYCLE_DIR\fR
.RS 4
Selects the window in the given (cyclic) direction\&.
.RE
.PP
biggest
.RS 4
Selects the biggest window on the current desktop\&.
.RE
.PP
last
.RS 4
Selects the previously focused window on the current desktop\&.
.RE
.PP
focused
.RS 4
Selects the currently focused window\&.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBModifiers\fR
.RS 4
.PP
floating
.RS 4
Only consider floating windows\&.
.RE
.PP
tiled
.RS 4
Only consider tiled windows\&.
.RE
.PP
like
.RS 4
Only consider windows that have the same class as the current window\&.
.RE
.PP
unlike
.RS 4
Only consider windows that have a different class than the current window\&.
.RE
.PP
automatic
.RS 4
Only consider windows in automatic splitting mode\&.
.RE
.PP
manual
.RS 4
Only consider windows in manual splitting mode (see
\fB\-\-presel\fR)\&.
.RE
.PP
urgent
.RS 4
Only consider urgent windows\&.
.RE
.PP
nonurgent
.RS 4
Only consider nonurgent windows\&.
.RE
.RE
.SS "Desktop"
.sp
Select a desktop\&.
.sp
.if n \{\
.RS 4
.\}
.nf
DESKTOP_SEL := <desktop_name>
             | ^<n>
             | (CYCLE_DIR|last|focused)[\&.occupied|\&.free][\&.urgent|\&.nonurgent]
.fi
.if n \{\
.RE
.\}
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBPrimary Selectors\fR
.RS 4
.PP
<desktop_name>
.RS 4
Selects the desktop with the given name\&.
.RE
.PP
^<n>
.RS 4
Selects the nth desktop\&.
.RE
.PP
\fICYCLE_DIR\fR
.RS 4
Selects the desktop in the given direction relative to the active desktop\&.
.RE
.PP
last
.RS 4
Selects the previously focused desktop\&.
.RE
.PP
focused
.RS 4
Selects the currently focused desktop\&.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBModifiers\fR
.RS 4
.PP
occupied
.RS 4
Only consider occupied desktops\&.
.RE
.PP
free
.RS 4
Only consider free desktops\&.
.RE
.PP
urgent
.RS 4
Only consider urgent desktops\&.
.RE
.PP
nonurgent
.RS 4
Only consider nonurgent desktops\&.
.RE
.RE
.SS "Monitor"
.sp
Select a monitor\&.
.sp
.if n \{\
.RS 4
.\}
.nf
MONITOR_SEL := <monitor_name>
             | ^<n>
             | (DIR|CYCLE_DIR|last|primary|focused)[\&.occupied|\&.free]
.fi
.if n \{\
.RE
.\}
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBPrimary Selectors\fR
.RS 4
.PP
<monitor_name>
.RS 4
Selects the monitor with the given name\&.
.RE
.PP
^<n>
.RS 4
Selects the nth monitor\&.
.RE
.PP
\fIDIR\fR
.RS 4
Selects the monitor in the given (spacial) direction relative to the active monitor\&.
.RE
.PP
\fICYCLE_DIR\fR
.RS 4
Selects the monitor in the given (cyclic) direction relative to the active monitor\&.
.RE
.PP
primary
.RS 4
Selects the primary monitor\&.
.RE
.PP
last
.RS 4
Selects the previously focused monitor\&.
.RE
.PP
focused
.RS 4
Selects the currently focused monitor\&.
.RE
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBModifiers\fR
.RS 4
.PP
occupied
.RS 4
Only consider monitors where the focused desktop is occupied\&.
.RE
.PP
free
.RS 4
Only consider monitors where the focused desktop is free\&.
.RE
.RE
.SH "COMMANDS"
.SS "Window"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
window [\fIWINDOW_SEL\fR] \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-f\fR, \fB\-\-focus\fR [\fIWINDOW_SEL\fR]
.RS 4
Focus the selected or given window\&.
.RE
.PP
\fB\-d\fR, \fB\-\-to\-desktop\fR \fIDESKTOP_SEL\fR
.RS 4
Send the selected window to the given desktop\&.
.RE
.PP
\fB\-m\fR, \fB\-\-to\-monitor\fR \fIMONITOR_SEL\fR
.RS 4
Send the selected window to the given monitor\&.
.RE
.PP
\fB\-w\fR, \fB\-\-to\-window\fR \fIWINDOW_SEL\fR
.RS 4
Transplant the selected window to the given window\&.
.RE
.PP
\fB\-s\fR, \fB\-\-swap\fR \fIWINDOW_SEL\fR
.RS 4
Swap the selected window with the given window\&.
.RE
.PP
\fB\-p\fR, \fB\-\-presel\fR \fIDIR\fR|cancel
.RS 4
Preselect the splitting area of the selected window (or cancel the preselection)\&.
.RE
.PP
\fB\-r\fR, \fB\-\-ratio\fR \fIRATIO\fR
.RS 4
Set the splitting ratio of the selected window (0 <
\fIRATIO\fR
< 1)\&.
.RE
.PP
\fB\-e\fR, \fB\-\-edge\fR \fIDIR\fR \fIRATIO\fR|pull|push
.RS 4
Set the splitting ratio (or pull, or push) the edge located in the given direction in relation to the selected window\&.
.RE
.PP
\fB\-R\fR, \fB\-\-rotate\fR \fIDIR\fR \fI90|270|180\fR
.RS 4
Rotate the tree holding the edge located in the given direction in relation to the selected window\&.
.RE
.PP
\fB\-t\fR, \fB\-\-toggle\fR floating|fullscreen|locked[=on|off]
.RS 4
Set or toggle the given state for the selected window\&.
.RE
.PP
\fB\-c\fR, \fB\-\-close\fR
.RS 4
Close the selected window\&.
.RE
.PP
\fB\-k\fR, \fB\-\-kill\fR
.RS 4
Kill the selected window\&.
.RE
.RE
.SS "Desktop"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
desktop [\fIDESKTOP_SEL\fR] \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-f\fR, \fB\-\-focus\fR [\fIDESKTOP_SEL\fR]
.RS 4
Focus the selected or given desktop\&.
.RE
.PP
\fB\-m\fR, \fB\-\-to\-monitor\fR \fIMONITOR_SEL\fR
.RS 4
Send the selected desktop to the given monitor\&.
.RE
.PP
\fB\-l\fR, \fB\-\-layout\fR \fICYCLE_DIR\fR|monocle|tiled
.RS 4
Set or cycle the layout of the selected desktop\&.
.RE
.PP
\fB\-n\fR, \fB\-\-rename\fR <new_name>
.RS 4
Rename the selected desktop\&.
.RE
.PP
\fB\-s\fR, \fB\-\-swap\fR \fIDESKTOP_SEL\fR
.RS 4
Swap the selected desktop with the given desktop\&.
.RE
.PP
\fB\-r\fR, \fB\-\-remove\fR
.RS 4
Remove the selected desktop\&.
.RE
.PP
\fB\-c\fR, \fB\-\-cancel\-presel\fR
.RS 4
Cancel the preselection of all the windows of the selected desktop\&.
.RE
.PP
\fB\-F\fR, \fB\-\-flip\fR \fIhorizontal|vertical\fR
.RS 4
Flip the tree of the selected desktop\&.
.RE
.PP
\fB\-R\fR, \fB\-\-rotate\fR \fI90|270|180\fR
.RS 4
Rotate the tree of the selected desktop\&.
.RE
.PP
\fB\-B\fR, \fB\-\-balance\fR
.RS 4
Adjust the split ratios of the tree of the selected desktop so that all windows occupy the same area\&.
.RE
.PP
\fB\-C\fR, \fB\-\-circulate\fR forward|backward
.RS 4
Circulate the leaves of the tree of the selected desktop\&.
.RE
.RE
.SS "Monitor"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
monitor [\fIMONITOR_SEL\fR] \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-f\fR, \fB\-\-focus\fR [\fIMONITOR_SEL\fR]
.RS 4
Focus the selected or given monitor\&.
.RE
.PP
\fB\-a\fR, \fB\-\-add\-desktops\fR <name>\&...
.RS 4
Create desktops with the given names in the selected monitor\&.
.RE
.PP
\fB\-r\fR, \fB\-\-remove\-desktops\fR <name>\&...
.RS 4
Remove desktops with the given names\&.
.RE
.PP
\fB\-n\fR, \fB\-\-rename\fR <new_name>
.RS 4
Rename the selected monitor\&.
.RE
.PP
\fB\-s\fR, \fB\-\-swap\fR \fIMONITOR_SEL\fR
.RS 4
Swap the selected monitor with the given monitor\&.
.RE
.RE
.SS "Query"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
query \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-W\fR, \fB\-\-windows\fR
.RS 4
List matching windows\&.
.RE
.PP
\fB\-D\fR, \fB\-\-desktops\fR
.RS 4
List matching desktops\&.
.RE
.PP
\fB\-M\fR, \fB\-\-monitors\fR
.RS 4
List matching monitors\&.
.RE
.PP
\fB\-T\fR, \fB\-\-tree\fR
.RS 4
Print tree rooted at query\&.
.RE
.PP
\fB\-H\fR, \fB\-\-history\fR
.RS 4
Print the history as it relates to the query\&.
.RE
.PP
\fB\-S\fR, \fB\-\-stack\fR
.RS 4
Print the window stacking order\&.
.RE
.PP
[\fB\-m\fR,\fB\-\-monitor\fR [\fIMONITOR_SEL\fR]] | [\fB\-d\fR,\fB\-\-desktop\fR [\fIDESKTOP_SEL\fR]] | [\fB\-w\fR, \fB\-\-window\fR [\fIWINDOW_SEL\fR]]
.RS 4
Constrain matches to the selected monitor, desktop or window\&.
.RE
.RE
.SS "Restore"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
restore \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-T\fR, \fB\-\-tree\fR <file_path>
.RS 4
Load the desktop trees from the given file\&.
.RE
.PP
\fB\-H\fR, \fB\-\-history\fR <file_path>
.RS 4
Load the focus history from the given file\&.
.RE
.PP
\fB\-S\fR, \fB\-\-stack\fR <file_path>
.RS 4
Load the window stacking order from the given file\&.
.RE
.RE
.SS "Control"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
control \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-\-adopt\-orphans\fR
.RS 4
Manage all the unmanaged windows remaining from a previous session\&.
.RE
.PP
\fB\-\-put\-status\fR
.RS 4
Write the current internal state to the panel FIFO\&.
.RE
.PP
\fB\-\-toggle\-visibility\fR
.RS 4
Toggle the visibility of all the windows\&.
.RE
.RE
.SS "Pointer"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
pointer \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-t\fR, \fB\-\-track\fR <x> <y>
.RS 4
Pass the pointer root coordinates for the current pointer action\&.
.RE
.PP
\fB\-g\fR, \fB\-\-grab\fR focus|move|resize_side|resize_corner
.RS 4
Perform the given pointer action\&.
.RE
.RE
.SS "Rule"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.sp
rule \fIOPTIONS\fR
.RE
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBOptions\fR
.RS 4
.PP
\fB\-a\fR, \fB\-\-add\fR <pattern> [\-d \fIDESKTOP_SEL\fR [\-\-follow]] [\-\-floating] [\-\-fullscreen] [\-\-locked] [\-\-focus] [\-\-unmanage] [\-\-one\-shot] [\-\-opacity <value>]
.RS 4
Create a new rule (<pattern> must match the class or instance name)\&.
.RE
.PP
\fB\-r\fR, \fB\-\-remove\fR <rule_uid>|tail|head\&...
.RS 4
Remove the rules with the given UIDs\&.
.RE
.PP
\fB\-l\fR, \fB\-\-list\fR [<pattern>]
.RS 4
List the rules\&.
.RE
.RE
.SS "Config"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.PP
config [\-m \fIMONITOR_SEL\fR|\-d \fIDESKTOP_SEL\fR] <key> [<value>]
.RS 4
Get or set the value of <key>\&.
.RE
.RE
.SS "Quit"
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBGeneral Syntax\fR
.RS 4
.PP
quit [<status>]
.RS 4
Quit with an optional exit status\&.
.RE
.RE
.SH "SETTINGS"
.sp

bspwm.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <ctype.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/select.h>
#include "types.h"
#include "desktop.h"
#include "monitor.h"
#include "settings.h"
#include "messages.h"
#include "events.h"
#include "common.h"
#include "bspwm.h"
#include "tree.h"
#include "window.h"
#include "history.h"
#include "stack.h"
#include "rule.h"
#include "ewmh.h"

int main(int argc, char *argv[])
{
    fd_set descriptors;
    char socket_path[MAXLEN];
    char *fifo_path = NULL;
    config_path[0] = '\0';
    status_prefix = NULL;
    int sock_fd, ret_fd, dpy_fd, sel, n;
    struct sockaddr_un sock_address;
    size_t rsp_len = 0;
    char msg[BUFSIZ] = {0};
    char rsp[BUFSIZ] = {0};
    xcb_generic_event_t *event;
    char opt;

    while ((opt = getopt(argc, argv, "hvc:s:p:")) != -1) {
        switch (opt) {
            case 'h':
                printf(WM_NAME " [-h|-v|-c CONFIG_PATH|-s PANEL_FIFO|-p PANEL_PREFIX]\n");
                exit(EXIT_SUCCESS);
                break;
            case 'v':
                printf("%s\n", VERSION);
                exit(EXIT_SUCCESS);
                break;
            case 'c':
                snprintf(config_path, sizeof(config_path), "%s", optarg);
                break;
            case 's':
                fifo_path = optarg;
                break;
            case 'p':
                status_prefix = optarg;
                break;
        }
    }

    if (config_path[0] == '\0') {
        char *config_home = getenv(CONFIG_HOME_ENV);
        if (config_home != NULL)
            snprintf(config_path, sizeof(config_path), "%s/%s/%s", config_home, WM_NAME, CONFIG_NAME);
        else
            snprintf(config_path, sizeof(config_path), "%s/%s/%s/%s", getenv("HOME"), ".config", WM_NAME, CONFIG_NAME);
    }

    dpy = xcb_connect(NULL, &default_screen);

    if (xcb_connection_has_error(dpy))
        err("Can't open the default display.\n");

    setup();

    dpy_fd = xcb_get_file_descriptor(dpy);

    char *sp = getenv(SOCKET_ENV_VAR);
    snprintf(socket_path, sizeof(socket_path), "%s", (sp == NULL ? DEFAULT_SOCKET_PATH : sp));

    sock_address.sun_family = AF_UNIX;
    snprintf(sock_address.sun_path, sizeof(sock_address.sun_path), "%s", socket_path);
    unlink(socket_path);

    sock_fd = socket(AF_UNIX, SOCK_STREAM, 0);

    if (sock_fd == -1)
        err("Couldn't create the socket.\n");

    if (bind(sock_fd, (struct sockaddr *) &sock_address, sizeof(sock_address)) == -1)
        err("Couldn't bind a name to the socket.\n");

    if (listen(sock_fd, SOMAXCONN) == -1)
        err("Couldn't listen to the socket.\n");

    sel = MAX(sock_fd, dpy_fd) + 1;

    if (fifo_path != NULL) {
        int fifo_fd = open(fifo_path, O_RDWR | O_NONBLOCK);
        if (fifo_fd != -1)
            status_fifo = fdopen(fifo_fd, "w");
        else
            warn("Couldn't open status fifo.\n");
    }

    load_settings();
    run_config();
    running = true;

    while (running) {

        xcb_flush(dpy);

        FD_ZERO(&descriptors);
        FD_SET(sock_fd, &descriptors);
        FD_SET(dpy_fd, &descriptors);

        if (select(sel, &descriptors, NULL, NULL, NULL) > 0) {

            if (FD_ISSET(sock_fd, &descriptors)) {
                ret_fd = accept(sock_fd, NULL, 0);
                if (ret_fd > 0 && (n = recv(ret_fd, msg, sizeof(msg), 0)) > 0) {
                    msg[n] = '\0';
                    if (handle_message(msg, n, rsp)) {
                        rsp_len = strlen(rsp);
                        if (rsp[rsp_len - 1] == '\n')
                            rsp[--rsp_len] = '\0';
                    } else {
                        rsp[0] = MESSAGE_FAILURE;
                        rsp_len = 1;
                    }
                    send(ret_fd, rsp, rsp_len, 0);
                    close(ret_fd);
                    rsp[0] = '\0';
                }
            }

            if (FD_ISSET(dpy_fd, &descriptors)) {
                while ((event = xcb_poll_for_event(dpy)) != NULL) {
                    handle_event(event);
                    free(event);
                }
            }

        }

        if (xcb_connection_has_error(dpy))
            err("The server has closed the connection.\n");
    }

    cleanup();
    close(sock_fd);
    if (status_fifo != NULL)
        fclose(status_fifo);
    xcb_ewmh_connection_wipe(ewmh);
    xcb_destroy_window(dpy, motion_recorder);
    free(ewmh);
    xcb_flush(dpy);
    xcb_disconnect(dpy);
    return exit_status;
}

void init(void)
{
    num_monitors = num_desktops = num_clients = 0;
    monitor_uid = desktop_uid = rule_uid = 0;
    mon = mon_head = mon_tail = pri_mon = NULL;
    rule_head = rule_tail = NULL;
    history_head = history_tail = NULL;
    stack_head = stack_tail = NULL;
    status_fifo = NULL;
    last_motion_time = last_motion_x = last_motion_y = 0;
    randr_base = 0;
    visible = auto_raise = true;
    exit_status = 0;
}

void setup(void)
{
    init();
    ewmh_init();
    screen = xcb_setup_roots_iterator(xcb_get_setup(dpy)).data;
    if (screen == NULL)
        err("Can't acquire the default screen.\n");
    root = screen->root;
    register_events();

    screen_width = screen->width_in_pixels;
    screen_height = screen->height_in_pixels;
    root_depth = screen->root_depth;

    uint32_t mask = XCB_CW_EVENT_MASK;
    uint32_t values[] = {XCB_EVENT_MASK_POINTER_MOTION};
    motion_recorder = xcb_generate_id(dpy);
    xcb_create_window(dpy, XCB_COPY_FROM_PARENT, motion_recorder, root, 0, 0, screen_width, screen_height, 0, XCB_WINDOW_CLASS_INPUT_ONLY, XCB_COPY_FROM_PARENT, mask, values);

    xcb_atom_t net_atoms[] = {ewmh->_NET_SUPPORTED,
                              ewmh->_NET_SUPPORTING_WM_CHECK,
                              ewmh->_NET_DESKTOP_NAMES,
                              ewmh->_NET_NUMBER_OF_DESKTOPS,
                              ewmh->_NET_CURRENT_DESKTOP,
                              ewmh->_NET_CLIENT_LIST,
                              ewmh->_NET_ACTIVE_WINDOW,
                              ewmh->_NET_WM_DESKTOP,
                              ewmh->_NET_WM_STATE,
                              ewmh->_NET_WM_STATE_FULLSCREEN,
                              ewmh->_NET_WM_STATE_DEMANDS_ATTENTION,
                              ewmh->_NET_WM_WINDOW_TYPE,
                              ewmh->_NET_WM_WINDOW_TYPE_DOCK,
                              ewmh->_NET_WM_WINDOW_TYPE_DESKTOP,
                              ewmh->_NET_WM_WINDOW_TYPE_NOTIFICATION,
                              ewmh->_NET_WM_WINDOW_TYPE_DIALOG,
                              ewmh->_NET_WM_WINDOW_TYPE_UTILITY,
                              ewmh->_NET_WM_WINDOW_TYPE_TOOLBAR};

    xcb_ewmh_set_supported(ewmh, default_screen, LENGTH(net_atoms), net_atoms);
    ewmh_set_supporting(motion_recorder);

#define GETATOM(a) \
    get_atom(#a, &a);
    GETATOM(WM_DELETE_WINDOW)
    GETATOM(WM_TAKE_FOCUS)
    GETATOM(_BSPWM_FLOATING_WINDOW)
    GETATOM(_NET_WM_WINDOW_OPACITY)
#undef GETATOM

    const xcb_query_extension_reply_t *qep = xcb_get_extension_data(dpy, &xcb_randr_id);
    if (qep->present && import_monitors()) {
        randr = true;
        randr_base = qep->first_event;
        xcb_randr_select_input(dpy, root, XCB_RANDR_NOTIFY_MASK_SCREEN_CHANGE);
    } else {
        randr = false;
        warn("Couldn't retrieve monitors via RandR.\n");
        xcb_rectangle_t rect = (xcb_rectangle_t) {0, 0, screen_width, screen_height};
        monitor_t *m = add_monitor(rect);
        add_desktop(m, make_desktop(NULL));
    }

    ewmh_update_number_of_desktops();
    ewmh_update_desktop_names();
    ewmh_update_current_desktop();
    frozen_pointer = make_pointer_state();
    xcb_get_input_focus_reply_t *ifo = xcb_get_input_focus_reply(dpy, xcb_get_input_focus(dpy), NULL);
    if (ifo != NULL && (ifo->focus == XCB_INPUT_FOCUS_POINTER_ROOT || ifo->focus == XCB_NONE))
        clear_input_focus();
    free(ifo);
}

void register_events(void)
{
    uint32_t values[] = {ROOT_EVENT_MASK};
    xcb_generic_error_t *e = xcb_request_check(dpy, xcb_change_window_attributes_checked(dpy, root, XCB_CW_EVENT_MASK, values));
    if (e != NULL) {
        xcb_disconnect(dpy);
        err("Another window manager is already running.\n");
    }
}

void quit(void)
{
    running = false;
}

void cleanup(void)
{
    while (mon_head != NULL)
        remove_monitor(mon_head);
    while (rule_head != NULL)
        remove_rule(rule_head);
    while (stack_head != NULL)
        remove_stack(stack_head);
    empty_history();
    free(frozen_pointer);
}

void put_status(void)
{
    if (status_fifo == NULL)
        return;
    if (status_prefix != NULL)
        fprintf(status_fifo, "%s", status_prefix);
    bool urgent = false;
    for (monitor_t *m = mon_head; m != NULL; m = m->next) {
        fprintf(status_fifo, "%c%s:", (mon == m ? 'M' : 'm'), m->name);
        for (desktop_t *d = m->desk_head; d != NULL; d = d->next, urgent = false) {
            for (node_t *n = first_extrema(d->root); n != NULL && !urgent; n = next_leaf(n, d->root))
                urgent |= n->client->urgent;
            char c = (urgent ? 'u' : (d->root == NULL ? 'f' : 'o'));
            if (m->desk == d)
                c = toupper(c);
            fprintf(status_fifo, "%c%s:", c, d->name);
        }
    }
    if (mon != NULL && mon->desk != NULL)
        fprintf(status_fifo, "L%s", (mon->desk->layout == LAYOUT_TILED ? "tiled" : "monocle"));
    fprintf(status_fifo, "\n");
    fflush(status_fifo);
}
